### PR TITLE
Added SkillAPI requirements when trying to purchase spells 

### PR DESCRIPTION
**Pull request by robotnikthingy**
Added SkillAPI requirements when trying to purchase spells, add options to check if player has required class or has researched the required skillAPI skill

I think I implemented it properly, let me know what you think

### CODE REVIEW

**Comment by NathanWolf** [comment]
Thanks! I will take a look at this as soon as I can, it seems like it’d be nice to have.

It’d be cool if you could squash this into one commit, but not a huge deal.

File: Magic/src/main/java/com/elmakers/mine/bukkit/action/BaseShopAction.java [hunk 1]
```diff
@@ -31,15 +29,11 @@
...
-import java.util.ArrayList;
-import java.util.Arrays;
-import java.util.Collection;
-import java.util.HashMap;
-import java.util.List;
-import java.util.Map;
-
 import javax.annotation.Nonnull;
 import javax.annotation.Nullable;
+import java.util.*;
```
Comment (killme) [inline Magic/src/main/java/com/elmakers/mine/bukkit/action/BaseShopAction.java]: Please don't contract imports.

File: Magic/src/main/java/com/elmakers/mine/bukkit/action/BaseShopAction.java [hunk 2]
```diff
@@ -186,7 +190,9 @@ public SpellResult checkContext(CastContext context) {
             }
         }
 
+        //ToDO: maybe add check for skillAPI here too, I think here is where they check to cast spells
         // Check path requirements
+
```
Comment (killme) [inline Magic/src/main/java/com/elmakers/mine/bukkit/action/BaseShopAction.java]: Please avoid any unrelated style changes.
It makes the diff more difficult to read.

File: Magic/src/main/java/com/elmakers/mine/bukkit/action/BaseShopAction.java [hunk 3]
```diff
@@ -532,6 +544,7 @@ public void prepare(CastContext context, ConfigurationSection parameters) {
         costScale = parameters.getDouble("scale", 1);
         filterBound = parameters.getBoolean("filter_bound", false);
         putInHand = parameters.getBoolean("put_in_hand", true);
+        //requiredSkillapiClass = parameters.getString("skillapi_class", null);
```
Comment (killme) [inline Magic/src/main/java/com/elmakers/mine/bukkit/action/BaseShopAction.java]: Commented out?
`#getString` returns null when nothing is available, so you can drop the `, null`
https://hub.spigotmc.org/javadocs/bukkit/org/bukkit/configuration/ConfigurationSection.html#getString-java.lang.String-

File: Magic/src/main/java/com/elmakers/mine/bukkit/action/builtin/SpellProgressAction.java [hunk 4]
```diff
@@ -99,6 +104,15 @@ public SpellResult perform(CastContext context) {
                     if (requiredPathKey != null && currentPath == null) {
                         continue;
                     }
+
+                    //SkillAPI
+                    if(requiredSkillapiClass != null && !mage.hasSkillAPIClass(requiredSkillapiClass)){
```
Comment (killme) [inline Magic/src/main/java/com/elmakers/mine/bukkit/action/builtin/SpellProgressAction.java]: `if(` -> `if (`
`){` -> `) {`

File: Magic/src/main/java/com/elmakers/mine/bukkit/action/builtin/SpellProgressAction.java [hunk 5]
```diff
@@ -99,6 +104,15 @@ public SpellResult perform(CastContext context) {
...
+                        lore.add(ChatColor.RED + context.getMessage("skillapi_class_requirement").replace("$skillapiclass", requiredSkillapiClass));
+                    }
+                    if(requiredSkillapiSkill != null && !mage.hasSkillAPISkill(requiredSkillapiSkill)){
```
Comment (killme) [inline Magic/src/main/java/com/elmakers/mine/bukkit/action/builtin/SpellProgressAction.java]: Same here

File: Magic/src/main/java/com/elmakers/mine/bukkit/action/builtin/SpellProgressAction.java [hunk 6]
```diff
@@ -99,6 +104,15 @@ public SpellResult perform(CastContext context) {
                     if (requiredPathKey != null && currentPath == null) {
                         continue;
                     }
+
+                    //SkillAPI
+                    if(requiredSkillapiClass != null && !mage.hasSkillAPIClass(requiredSkillapiClass)){
+                        lore.add(ChatColor.RED + context.getMessage("skillapi_class_requirement").replace("$skillapiclass", requiredSkillapiClass));
```
Comment (killme) [inline Magic/src/main/java/com/elmakers/mine/bukkit/action/builtin/SpellProgressAction.java]: @NathanWolf The color should be in the language file right?

File: Magic/src/main/java/com/elmakers/mine/bukkit/integration/SkillAPIManager.java [hunk 7]
```diff
@@ -94,4 +93,20 @@ public void setMana(Player player, float amount) {
...
+    }
+
+    @Override
+    public boolean hasClass(Player player, String name){
+        PlayerData playerData = SkillAPI.getPlayerData(player);
+
+        return playerData.hasClass(name);
+    }
+
+
```
Comment (killme) [inline Magic/src/main/java/com/elmakers/mine/bukkit/integration/SkillAPIManager.java]: Double blank line

File: Magic/src/main/java/com/elmakers/mine/bukkit/magic/Mage.java [hunk 8]
```diff
@@ -75,23 +68,12 @@
...
-import com.elmakers.mine.bukkit.block.MaterialBrush;
-import com.elmakers.mine.bukkit.block.UndoQueue;
-import com.elmakers.mine.bukkit.batch.UndoBatch;
-import com.elmakers.mine.bukkit.wand.Wand;
-
 import javax.annotation.Nonnull;
 import javax.annotation.Nullable;
+import java.lang.ref.WeakReference;
+import java.lang.reflect.Method;
+import java.util.*;
```
Comment (killme) [inline Magic/src/main/java/com/elmakers/mine/bukkit/magic/Mage.java]: imports

File: Magic/src/main/java/com/elmakers/mine/bukkit/spell/BaseSpell.java [hunk 9]
```diff
@@ -82,7 +56,7 @@
     protected static final int MIN_Y = 1;
 
     // TODO: Configurable default? this does look cool, though.
-    protected final static Material DEFAULT_EFFECT_MATERIAL = Material.WATER;
+    protected final static Material DEFAULT_EFFECT_MATERIAL = Material.STATIONARY_WATER;
```
Comment (killme) [inline Magic/src/main/java/com/elmakers/mine/bukkit/spell/BaseSpell.java]: Why?

File: .gitignore [hunk 10]
```diff
@@ -7,3 +7,4 @@ src/main/resources/spells.defaults.yml
 src/main/resources/wands.defaults.yml
 **/original_skin.png
 **/mc_creds.php
+*.project
```
Comment (killme) [inline .gitignore]: Shouldn't this be just `.project`, as you want to get rid of all files that are named exactly that.

File: Magic/src/main/java/com/elmakers/mine/bukkit/action/builtin/SpellProgressAction.java (hunk 6 above)
Comment (robotnikthingy) [reply Magic/src/main/java/com/elmakers/mine/bukkit/action/builtin/SpellProgressAction.java]: Im not sure, if you look a few lines below it though for checking the required amount of casts, it does the same thing. I just decided to follow that

```
if (castCount == requiredCastCount) {
    lore.add(ChatColor.GREEN + context.getMessage("cast_requirement")
            .replace("$current", Long.toString(castCount))
            .replace("$required", Long.toString(requiredCastCount)));
} else {
    lore.add(ChatColor.RED + context.getMessage("cast_requirement")
            .replace("$current", Long.toString(castCount))
            .replace("$required", Long.toString(requiredCastCount)));
}
```

File: Magic/src/main/java/com/elmakers/mine/bukkit/spell/BaseSpell.java (hunk 9 above)
Comment (robotnikthingy) [reply Magic/src/main/java/com/elmakers/mine/bukkit/spell/BaseSpell.java]: Not sure what happened there, I dont even remember changing it

**Comment by robotnikthingy** [comment]
I've made the requested changes. If there is anything else needed let me know.

**Comment by NathanWolf** [comment]
There are definitely some weird things with this PR I will need to look at more closely. Some of them might just be merge issues, but also:

- We can't replace the preconditions without breaking backwards compatibility with older version of Java (maybe and/or older Spigot versions?)

- I agree with @killme that it'd be nice if this were less SkillAPI-specific. "Required class" for instance would be a great generalized fit for spells in the spell shop given that Magic now has its own class system, maybe we can tie in SkillAPI classes to that. I need to take a closer look though.

- I'm not sure how or if the above applies to "require skill" requirements, I don't think that necessarily applies to anything internal in Magic but maybe as @killme said there's an opportunity to expand on that idea.

- There seem to be a lot of unrelated changes, it'd be good to minimize these diffs. Re-arranging imports, whitespace changes, etc are not ideal.

So at this point I am leaning towards maybe taking pieces of this (like the expanded SkillAPIManager functionality) but perhaps rethinking some of the core integration.

I'm going to be a little tied up this weekend though, I want to take a closer look at this when I have some more time to dedicate. 

Thank you!

**Comment by robotnikthingy** [comment]
I had forgotten all about the builtin class system, probably because its still kinda new. The main reason I'm using SkillAPI on my server right now is for the class system just about only that, but it would be great if at some point I could ditch that and move to using Magic's built in class system. Not sure if Magic's class system has it yet, but an attribute system similar to SkillAPI would be a great idea. With that being said, making it so that Magic classes are taken into account too would obviously be a great idea.

As for the SkillAPI skills, I was planning to use the skills at some point as a part of a progression system, where a player has to research a certain skill before they can purchase a spell. I was also looking into doing the same thing with attributes too, such as requiring the player to have a certain level of intelligence

It might be worth looking into having a requirement system for purchasing spells or magic item, to make adding new requirements easier. Right now I think we can just check for if players have the required spell points or path, adding more requirements would allow servers to build more complex and customization skill trees, rather than just following the linear path system.

**Comment by NathanWolf** [comment]
I definitely like the idea of modularizing the requirements. I would like to reuse this for the new Selector action, too.

A native attribute system is definitely on my near term target list. Check ROADMAP.md if you’re interested!

I can’t think of a good parallel for “require skill” in native magic, I guess a spell would be the closest so that doesn’t seem like a good fit. Abstracting requirements is probably the best way to go with all of this.

**Comment by NathanWolf** [comment]
I have merged this into a feature branch, but I plan on picking it apart quite a bit before putting it in master. 

Reviewing the code, I'm not sure this works though? It seems like all it's going to do is add requirements lore in the spell progress menu, but it doesn't actually enforce the requirements anywhere?

What I am planning on doing:
1. Take your changes to SkillAPIManager and maybe AttributeProvider (not really sure skills/classes belong there.. probably not)
2. Add requirements to SpellTemplate .. not sure yet if I'm going to add separate "requiredSkill" and "requiredClass" properties or get more industrious and try to make some more generic requirements system. @killme any thoughts?
3. Enforce requirements in SpellShopAction so the spells don't show up.
4. Maybe also add an option to SpellShopAction to show unavailable spells - this has been requested for other requirements, too, I think. Could use the disabled spell icons, which would work out nicely.

Let me know if either of you have further thoughts here, I'd like to get this wrapped up before the next release.

**Comment by killme** [comment]
I think ideally we'd have the ability to specify predicates similar to advancements (but without the triggering behaviors). I think we could probably elsewhere as well. But I am not sure if this worth the effort. As at AMC we might end up writing something custom eventually anyways.

Example of how we have implemented such predicates elsewhere at amc:
```lang=json
  "visible": {
    "criteria": {
      "in_fnc": {
        "type": "in_nav_region",
        "region": "fnc/region"
      },
      "air/lvl_8": {
        "type": "advancement_done",
        "advancement": "amc:air/reach_level_8"
      },
      "water/lvl_8": {
        "type": "advancement_done",
        "advancement": "amc:water/reach_level_8"
      },
      "earth/lvl_8": {
        "type": "advancement_done",
        "advancement": "amc:earth/reach_level_8"
      },
      "fire/lvl_8": {
        "type": "advancement_done",
        "advancement": "amc:fire/reach_level_8"
      }
    },
    "requirements": [
      [ "in_fnc" ],
      [ "air/lvl_8", "water/lvl_8", "earth/lvl_8", "fire/lvl_8" ]
    ]
  }
```

But that is of course a completely different style of doing configurations.

However, if we do implement predicates, we may be able to reuse some of it in other spells that might need it for some reason?

**Comment by NathanWolf** [comment]
Since you already have such a complex system implemented, it seems like what would be the most beneficial would be if you could tag spells with arbitrary requirements that would map into your own system?

I mean versus me trying to make something that complex in Magic, which would ultimately be redundant for you anyway.

So something like 
```
requirements:
  - class:somemagicclass
  - skillapiclass:someskillapiclass
  - skillapiskill:someskill
  - amc:air/level_8
```

And then you could register a requirement processor that basically gets this list of requirements and says if a player has them or not? For your processor you'd only care about the "amc" prefixed ones, and internally this would give a nice separation for skillapi requirements and anything else we might add in the future.

The processor could maybe also be responsible for returning some text to add to lore, so you could say like "Requires Airbending Level 8"?

Maybe instead of prefixes it could be like a type/name combo, a little more awkward to configure but cleaner to implement?

**Comment by killme** [comment]
That would work. Ideally you also support parameters, so you it will look slightly better.
```yaml
requirements:
- type: skillapi
  class: something
```

**Comment by NathanWolf** [comment]
Support parameters meaning you'd like your extension to get that whole config block? So in theory you could config something like

```
requirements:
 - type: amc
    class: airbending
    level: 8
    quest: openspiritportal
    item: airstaff
```

etc?

So if we were to make an interface,  maybe something like
```
public interface RequirementProcessor {
  // Return null if you don't care about this requirement type or if the player passes
  // Return a description of the requirement if the player does not pass
  String checkRequirement(String type, ConfigurationSection configuration);
```

Thoughts on that?

**Comment by killme** [comment]
I think you should let it return a new interface, the configured processor.
Which then has a `test(Player)` and `chat getRequirementDisplay(Player)` for testing if the player matches, and for rendering a description line..

**Comment by NathanWolf** [comment]
Oh I think you are a step ahead of me

I wasn't thinking of adding a centralized definition of requirements

All those configs I've typing I assumed you would put them right on a spell. (or in the future whatever other thing could have requirements)

So like if you wanted to lock out Air Blast in some special way you'd put it on the spell definition directly, like
```
airblast:
  requirements:
  - type: amc
     advancement: amc:air/reach_level_8
```

Etc

Then on your side you just register a single RequirementProcessor that handles any of them.

Does that sound useable? Otherwise I'll have to think more about what you said, having to preconfigure them all seems like more of a pain but not sure.

**Comment by NathanWolf** [comment]
Also worth noting is the only places I would initially support requirements would be on spells, used by the spellshop - and maybe on Selector action options, too.

**Comment by dumptruckman** [comment]
Sounds like a good plan. I do wonder if more params are needed in the callback. I'd say just make sure that the existing stuff can be converted to the new system.

**Comment by dumptruckman** [comment]
The interface should probably have a method to return the format for displaying the requirements in the shop.

**Comment by NathanWolf** [comment]
Feedback please?

https://github.com/elBukkit/MagicPlugin/blob/master/MagicAPI/src/main/java/com/elmakers/mine/bukkit/api/requirements/RequirementsProcessor.java

**Comment by dumptruckman** [comment]
Not sure I understand the requirementType string. Can you give an example? Also, would it make more sense to pass in Mage rather than Player?

**Comment by NathanWolf** [comment]
Yeah mage probably makes more sense.

So for type that'd be like "skillapi" or "avatarmc" or "mycoolpluginwithrequirements".

Actually ... maybe it would be better to require that you register the type when you register your RequirementProcessor?

It's sort of meant to be like a domain, so a RequirementProcessor would only care about requirements of a certain type.

**Comment by NathanWolf** [comment]
Better?

https://github.com/elBukkit/MagicPlugin/commit/a778cdea988bc71eac84ed5360f16e9d246ce47d

**Comment by dumptruckman** [comment]
Yeah, I think I like the registration idea better.

**Comment by NathanWolf** [comment]
I pushed what I'm hoping is the finalized API, if anyone cares to take a look. It's basically what I last posted, except I'm passing a CastContext along instead of a Mage.

I've got SkillAPI support implemented in SpellShop, working on Selector and then I'll add a native provider for the type "magic" (which is the default type). I'm going to move over all of the stuff I copied from shops into Selector, like requiring a wand, class, path, etc.

@robotnikthingy I think this satisfies the intent of your PR, you can now add skill API class and skill requirements to acquiring spells.

**Comment by robotnikthingy** [comment]
Sounds great, looking forward to the next update!
